x/subscription/keeper/subscription.go

```diff
@@ -401,6 +407,121 @@ func (k Keeper) resetSubscriptionDetailsAndAppendEntry(ctx sdk.Context, sub *typ
+			utils.LogAttr("creator", creator),
+			utils.LogAttr("price", price),
+		)
+	}
+
+	return nil
+}
+
+func (k Keeper) CreateFutureSubscription(ctx sdk.Context,
+	creator string,
+	consumer string,
+	planIndex string,
+	duration uint64,
+) error {
+	var err error
+
+	block := uint64(ctx.BlockHeight())
+	creatorAcct, plan, err := k.verifySubscriptionBuyInput(ctx, block, creator, consumer, planIndex)
+	if err != nil {
+		return err
+	}
+
+	var sub types.Subscription
+	found := k.subsFS.FindEntry(ctx, consumer, block, &sub)
+
+	if !found {
+		return utils.LavaFormatWarning("could not find active subscription. advanced purchase is only available for an active subscription", nil,
+			utils.LogAttr("consumer", consumer),
+			utils.LogAttr("block", block),
+		)
+	}
+
+	createFutureSubscription := func(sub *types.Subscription, creator, planIndex string, planBlock, duration uint64) {
+		sub.FutureSubscription = &types.FutureSubscription{
+			Creator:        creator,
+			PlanIndex:      planIndex,
+			PlanBlock:      planBlock,
+			DurationBought: duration,
+		}
+	}
+
+	if sub.FutureSubscription != nil {
+		// Consumer already has a future subscription
+		// If the new plan's price > current future subscription's plan - change and charge the diff
+		currentPlan, err := k.GetPlanFromSubscription(ctx, consumer, block)
+		if err != nil {
+			return utils.LavaFormatError("panic: could not get active subscription's plan. aborting", err,
+				utils.Attribute{Key: "creator", Value: creator},
+			)
+		}
+
+		newPlanCostForDuration := plan.Price.Amount.MulRaw(int64(duration))
+		consumerPaid := currentPlan.Price.Amount.MulRaw(int64(sub.FutureSubscription.DurationBought))
+		if newPlanCostForDuration.GT(consumerPaid) {
+			priceDiff := sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), newPlanCostForDuration.Sub(consumerPaid))
+			err = k.chargeFromCreatorAccountToModule(ctx, creatorAcct, priceDiff)
+			if err != nil {
+				return err
+			}
+
+			createFutureSubscription(&sub, creator, plan.Index, plan.Block, duration)
+
+			details := map[string]string{
+				"creator":      creator,
+				"consumer":     consumer,
+				"duration":     strconv.FormatUint(duration, 10),
+				"oldPlanIndex": currentPlan.Index,
+				"oldPlanBlock": strconv.FormatUint(currentPlan.Block, 10),
+				"newPlanIndex": plan.Index,
+				"newPlanBlock": strconv.FormatUint(plan.Block, 10),
+			}
+			utils.LogLavaEvent(ctx, k.Logger(ctx), types.AdvancedBuyUpgradeSubscriptionEventName, details, "advanced subscription upgraded")
+			return nil
+		} else {
+			return utils.LavaFormatWarning("can't purchase another plan in advanced with a lower price", nil)
+		}
+	}
+
+	price := plan.GetPrice()
+	price.Amount = price.Amount.MulRaw(int64(duration))
+	k.applyPlanDiscountIfEligible(duration, &plan, &price)
+
+	err = k.chargeFromCreatorAccountToModule(ctx, creatorAcct, price)
+	if err != nil {
+		return err
+	}
+
+	createFutureSubscription(&sub, creator, plan.Index, plan.Block, duration)
```

Comment:
why do you need this if the input is the same in both places

Reply:
Correct. Fixed in 681d0cc5066e62640fea0447e4234bc160692d82